resolver/resolver.cc

```diff
@@ -553,46 +553,69 @@ class ResolveConstantsWalk {
+                if (auto e =
+                        gs.beginError(core::Loc(todo.file, id->loc), core::errors::Resolver::InvalidMixinDeclaration)) {
+                    e.setHeader("Argument to `{}` must be statically resolvable to a module", send->fun.show(gs));
+                }
+                return;
             }
-            return;
-        }
-        if (id->symbol == owner) {
-            if (auto e =
-                    gs.beginError(core::Loc(todo.file, send->loc), core::errors::Resolver::InvalidMixinDeclaration)) {
-                e.setHeader("Must not pass your self to `{}`", send->fun.show(gs));
+            if (id->symbol.data(gs)->isClassOrModuleClass()) {
+                if (auto e =
+                        gs.beginError(core::Loc(todo.file, id->loc), core::errors::Resolver::InvalidMixinDeclaration)) {
+                    e.setHeader("`{}` is a class, not a module; Only modules may be mixins",
+                                id->symbol.data(gs)->show(gs));
+                }
+                encounteredError = true;
             }
-            return;
-        }
-        auto existing = owner.data(gs)->findMember(gs, core::Names::classMethods());
-        if (existing.exists() && existing != id->symbol) {
-            if (auto e =
-                    gs.beginError(core::Loc(todo.file, send->loc), core::errors::Resolver::InvalidMixinDeclaration)) {
-                e.setHeader("Redeclaring `{}` from module `{}` to module `{}`", send->fun.show(gs),
-                            existing.data(gs)->show(gs), id->symbol.data(gs)->show(gs));
+            if (id->symbol == owner) {
+                if (auto e =
+                        gs.beginError(core::Loc(todo.file, id->loc), core::errors::Resolver::InvalidMixinDeclaration)) {
+                    e.setHeader("Must not pass your self to `{}`", send->fun.show(gs));
+                }
+                encounteredError = true;
             }
-            return;
+            if (encounteredError) {
+                continue;
+            }
+
+            // Get the fake property holding the mixes
+            auto mixMethod = owner.data(gs)->findMember(gs, core::Names::mixedInClassMethods());
+            auto loc = core::Loc(owner.data(gs)->loc().file(), send->loc);
+            if (!mixMethod.exists()) {
+                // We never stored a mixin in this symbol
+                // Create a the fake property that will hold the mixed in modules
+                mixMethod = gs.enterMethodSymbol(loc, owner, core::Names::mixedInClassMethods());
+                vector<core::TypePtr> targs;
+                mixMethod.data(gs)->resultType = core::TupleType::build(gs, move(targs));
+
+                // Create a dummy block argument to satisfy sanitycheck during GlobalState::expandNames
+                auto &arg = gs.enterMethodArgumentSymbol(core::Loc::none(), mixMethod, core::Names::blkArg());
+                arg.flags.isBlock = true;
```

Comment:
@elliottt Ended up abandoning the previous implementation of adding arguments since it wasn't optimal as you mentioned and I couldn't figure out resolving the bug.

We are now using `TupleType` to store the modules (arguments). However, there was a sanity check during compilation that threw us off and ended up adding this dummy `<blk>` argument for now. 

Here's the stacktrace. This fake property (`<mixed_in_class_methods>`) doesn't have an argument and it fails the `Symbol::sanityCheck()`.

```
ERROR: /Users/kaanozkan/dev/forked/sorbet/payload/binary/BUILD:71:8: Executing genrule //payload/binary:generate-state-payload-raw failed (Exit 1)
Exception::raise(): core/Symbols.cc:1257 enforced condition !this->arguments().empty() has failed: T::Props#<mixed_in_class_methods>

Backtrace:
sorbet::core::Symbol::sanityCheck(sorbet::core::GlobalState const&) const (in sorbet-orig) + 1484
sorbet::core::GlobalState::sanityCheck() const (in sorbet-orig) + 947
sorbet::core::GlobalState::expandNames(unsigned int) (in sorbet-orig) + 28
sorbet::core::GlobalState::preallocateTables(unsigned int, unsigned int, unsigned int, unsigned int, unsigned int, unsigned int) (in sorbet-orig) + 200
sorbet::realmain::realmain(int, char**) (in sorbet-orig) + 2167
main (in sorbet-orig) + 11
start (in libdyld.dylib) + 1

Target //main:sorbet failed to build
```

Reply:
That makes sense -- the namer also adds the block argument for the fake method that stores the sealed subclasses.

Reply:
I'm curious as to why this is happening. Looking at https://github.com/sorbet/sorbet/blob/master/docs/pipeline.md I can see the `resolver` runs after `namer` and I'd assume preallocation of `Name`, `Symbol` tables would be done earlier than the `resolver` 🤔

Reply:
In debug builds we run the sanity checker after the resolver. This line is checking that the method has arguments present, as every method should at least have a block parameter. You weren't triggering it before, because you were storing the symbols in the arguments list, causing that sanity check to pass :)

Reply:
Sorry, I meant to include this link when referring to "this line":
https://github.com/sorbet/sorbet/blob/master/core/Symbols.cc#L1257